### ISSUE TITLE
Job.fetch_many() API frequently fails with NoSuchJobError

### ISSUE THREAD

**Issue by emadelwany** [rq/rq#1575]
Using version 1.8.1

Stack Trae:

```
File "/opt/conda/envs/ann38/lib/python3.8/site-packages/rq/job.py", line 329, in fetch_many
job.restore(results[i])
File "/opt/conda/envs/ann38/lib/python3.8/site-packages/rq/job.py", line 484, in restore
raise NoSuchJobError('Unexpected job format: {0}'.format(obj))
rq.exceptions.NoSuchJobError: Unexpected job format: {'last_heartbeat': b'2021-06-15T01:34:42.185285Z'}
```

**Comment by lowercase00** [comment]
@emadelwany do you have a sample to replicate this?

**Comment by emadelwany** [comment]
@lowercase00 -- unfortunately we were never able to pin point the issue, it's sporadic so it seems like a race condition. Completely speculating based on the error, but is it possible that some control plane job (heart beat) is interfering with data plane objects (our actual jobs)?

Any way, if it's helpful, we got around it by replacing this code:
```
   def get_jobs_from_registry(registry):
        return [job_to_dict(job) for job in Job.fetch_many(registry.get_job_ids(), connection=client) if job]
```

with this code:
```
def get_jobs_from_registry(registry: BaseRegistry) -> List[Dict[str, Any]]:
        job_ids: List[str] = registry.get_job_ids()
        jobs: List[Job]
        try:
            jobs = Job.fetch_many(job_ids, connection=client)
        # This is a workaround for this bug in RQ: https://github.com/rq/rq/issues/1575
        # When this gets resolved, we can go back to using fetch_many() exclusively and get rid of the error handling
        except NoSuchJobError:
            logging.exception("Failed to fetch jobs using bulk API, will fallback to non-bulk API")
            jobs = []
            for job_id in job_ids:
                try:
                    job = Job.fetch(job_id, connection=client)
                    if job is not None:
                        jobs.append(job)
                except NoSuchJobError:
                    logging.exception("Failed to fetch job")
                    continue

        return [job_to_dict(job) for job in jobs if job]
```

Which does the job albeit a bit slower.

**Comment by lowercase00** [comment]
What you describe makes sense: in fetch_many, an error will raise, but you won't be able to recover the rest (it's all or nothing). on the fetch one, it will still raise, but you are catching the exception and just continuing. Those two functions basically do the same thing, albeit the fetch many runs in a pipeline.

The question is why it fails in the first place, and that's basically because its missing the "data" key on the job object (thats where RQ stores the function to call, args and kwargs). 

- Are you using a default serializer?
- Python version
- Redis Py version
- Redis Version

One thing that would help understand is: log the job id that failed (on the except block), and then check (1) the job in Redis, and (2) the actual place where you enqueued the job to check to see if there’s anything different.

@selwin, I think we need a more aggressive debug logging strategy for these cases.

**Comment by selwin** [comment]
I think the bug @emadelwany ran into is already fixed.

This error `rq.exceptions.NoSuchJobError: Unexpected job format: {'last_heartbeat': b'2021-06-15T01:34:42.185285Z'}` was caused by a race condition on result handling.

I'm open to modifying the behavior of `fetch_many()` to return `None` if there are some jobs that can't be fetched for some reason.

**Comment by lowercase00** [comment]
We can have a param to decide whether is all or nothing, or if it should just ignore exceptions while retrieving specific jobs. Also improve the logging would be helpful.

**Comment by selwin** [comment]
Yes, the default should be the current behavior so we don't introduce any backward compatibility.

**Comment by emadelwany** [comment]
@lowercase00 I checked our production logs over the past 4 weeks and don't see the error getting logged, so it's possible it got resolved as @selwin said, we're currently on version 1.10.0 of the library. 

I think it's safe to close this issue, I'll re-open if we encounter it again. Thanks both for looking into it and your work maintaining this awesome OSS.

**Comment by lowercase00** [comment]
@emadelwany thanks for the follow up and closing this. We'll look into introducing a more obvious way of dealing with this scenario in the future, any other issues let us know.

**Comment by OlegZv** [comment]
I wonder if this should be reopened? I just hit similar issue, but in my case the redis state was `rq.exceptions.NoSuchJobError: Unexpected job format: {'status': b'canceled'}`
I'll use the @emadelwany 's workaroud for now, but curious if there's a more robus way to handle it.

**Comment by selwin** [comment]
What do you think should happen when Kob.fetch finds a key with unexpected data?

**Comment by OlegZv** [comment]
I didn't get a chance yet to investigate further. I don't know for sure, but it seems that the job does exist, but in one of the sates where there's no "data" key. My first hunch is to have these "special" states as another way to initialize the job. This might not be elegant, since there might be many of these states. I'll respond with more constructive thoughts this weekend.
